pkg/usagestats/reporter_test.go

```diff
@@ -0,0 +1,303 @@
+				cortex_usage_stats_report_sends_total 1
+
+				# HELP cortex_usage_stats_report_sends_failed_total The total number of failed send requests.
+				# TYPE cortex_usage_stats_report_sends_failed_total counter
+				cortex_usage_stats_report_sends_failed_total 1
+			`), "cortex_usage_stats_report_sends_total", "cortex_usage_stats_report_sends_failed_total"))
+		},
+	}
+
+	for testName, testRun := range tests {
+		t.Run(testName, func(t *testing.T) {
+			// Reset.
+			serverInvoked.Store(false)
+
+			testRun(t)
+		})
+	}
+}
+
+func TestReporter_SendReportPeriodically(t *testing.T) {
+	tests := map[string]struct {
+		failOneRequestEvery int
+	}{
+		"all requests to stats server succeed": {
+			failOneRequestEvery: 0,
+		},
+		"some requests to stats server fail and they get retried": {
+			failOneRequestEvery: 2,
+		},
+	}
+
+	for testName, testData := range tests {
+		testData := testData
+
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			var (
+				ctx           = context.Background()
+				bucketClient  = prepareLocalBucketClient(t)
+				reportsMx     sync.Mutex
+				reports       []Report
+				requestsCount = atomic.NewInt32(0)
+			)
+
+			// Mock the stats server.
+			server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+				requestsCount.Inc()
+				if testData.failOneRequestEvery > 0 && requestsCount.Load()%int32(testData.failOneRequestEvery) == 0 {
+					writer.WriteHeader(http.StatusServiceUnavailable)
+					return
+				}
+
+				data, err := io.ReadAll(request.Body)
+				require.NoError(t, err)
+
+				report := Report{}
+				require.NoError(t, json.Unmarshal(data, &report))
+
+				reportsMx.Lock()
+				reports = append(reports, report)
+				reportsMx.Unlock()
+
+				writer.WriteHeader(http.StatusOK)
+			}))
+			defer server.Close()
+
+			r := NewReporter(bucketClient, log.NewNopLogger(), prometheus.NewPedanticRegistry())
+			r.serverURL = server.URL
+			r.reportCheckInterval = 100 * time.Millisecond
+			r.reportSendInterval = time.Second
+
+			// Upload the seed file.
+			seed := newClusterSeed()
+			seed.CreatedAt = time.Now().Add(-time.Hour)
+			require.NoError(t, writeSeedFile(ctx, r.bucket, seed))
+
+			// Start the reporter.
+			require.NoError(t, services.StartAndAwaitRunning(ctx, r))
+			t.Cleanup(func() {
+				require.NoError(t, services.StopAndAwaitTerminated(ctx, r))
+			})
+
+			// We expect to have received a report per second.
+			test.Poll(t, 10*time.Second, true, func() interface{} {
+				reportsMx.Lock()
+				defer reportsMx.Unlock()
+				return len(reports) >= 4
+			})
+
+			reportsMx.Lock()
+			defer reportsMx.Unlock()
+
+			// We expect each report interval to be exactly at 1s apart from the previous one.
+			for i := 1; i < len(reports); i++ {
+				require.Equal(t, reports[i-1].Interval.Add(time.Second), reports[i].Interval)
+			}
+		})
+	}
+}
+
+func TestReporter_SendReportShouldSkipToNextReportOnLongFailure(t *testing.T) {
+	t.Parallel()
```

Comment:
AFAIK this does nothing here.

Reply:
It depends if there are tests with `t.Parallel()`. For example the two following tests can be executed in parallel:

```
func TestOne(t *testing.T) {
	t.Parallel()

	// ...
}

func TestTwo(t *testing.T) {
	t.Parallel()

	// ...
}
```

I generally prefer to flag a slow test which is safe to run parallelly with `t.Parallel()` without worrying if there are other parallel tests. If in the future we'll add them, they will run in parallel.